### PR TITLE
feat(signed-vouches): add ability to vouch using signatures

### PR DESCRIPTION
**Pull request by nix1g**
The EIP-712 domain is:
```js
EIPDomain {
    name: 'Proof of Humanity',
    chainId: block.chainid,
    verifyingContract: address(this)
}
```

The message expected to be signed is:
```solidity
struct IsHumanVoucher {
    address vouchedSubmission;
    uint256 voucherExpirationTimestamp;
}
```

### CODE REVIEW

File: contracts/ProofOfHumanity.sol [hunk 1]
```diff
@@ -469,15 +469,52 @@ contract ProofOfHumanity is IArbitrable, IEvidence {
...
+
+            Submission storage voucher = submissions[voucherAddress];
+            if (!voucher.hasVouched && voucher.registered && now - voucher.submissionTime <= submissionDuration &&
+            now < _timeouts[i] && _submissionID != voucherAddress) {
+                request.vouches.push(voucherAddress);
+                voucher.hasVouched = true;
+            }
+        }
+
         for (uint i = 0; i<_vouches.length && request.vouches.length<requiredNumberOfVouches; i++) {
```
Comment (epiqueras) [inline contracts/ProofOfHumanity.sol:518]: If `requiredNumberOfVouches > _signatures.length`, `request.vouches` will have unnecessary extra vouches.

Comment (nix1g) [reply contracts/ProofOfHumanity.sol:518]: There is request.vouches.length < requiredNumberOfVouches in both loops, how could this happen?

Comment (epiqueras) [reply contracts/ProofOfHumanity.sol:518]: Ahh, missed that.

**Review by ferittuncer — COMMENTED** [review]
Misleading variable name. Implies the time when renewal happened but actually it is the earliest time you can apply for renewal. I suggest renewalAvailableAt instead.


https://github.com/Proof-Of-Humanity/Proof-Of-Humanity/blob/18c66ae1036d700e11b418944aeb542623441210/contracts/ProofOfHumanity.sol#L440


What is the procedure for selecting a prefix? 

https://github.com/Proof-Of-Humanity/Proof-Of-Humanity/blob/18c66ae1036d700e11b418944aeb542623441210/contracts/ProofOfHumanity.sol#L518 



In changeStateToPending signature length is not checked. This might be a crypto flaw.

In changeStateToPending I think we have to compare the original address and recovered address to complete cryptographic verification. This might be crypto flaw: https://github.com/Proof-Of-Humanity/Proof-Of-Humanity/blob/18c66ae1036d700e11b418944aeb542623441210/contracts/ProofOfHumanity.sol#L541


A tutorial about signatures I followed during review: https://yos.io/2018/11/16/ethereum-signatures/

**Comment by nix1g** [comment]
> What is the procedure for selecting a prefix?
This is [EIP-712 specification](https://eips.ethereum.org/EIPS/eip-712), complying with [EIP-191](https://eips.ethereum.org/EIPS/eip-191). This is used to disambiguate between signed transaction (which are RLP-encoded and don't start with "\x19") and regular signed messages (which start with "\x19Ethereum Signed Message:")

> In changeStateToPending signature length is not checked. This might be a crypto flaw.
There is no harm providing an invalid signature, the recovered address will simply be wrong

> In changeStateToPending I think we have to compare the original address and recovered address to complete cryptographic verification. This might be a crypto flaw.
No since the caller already has to provide everything necessary to know the original address:
- its own public address (that is being vouched for)
- the expiration time of the signature
- the signature itself

**Pull request merged**